### ISSUE TITLE
how can i  finetune on your published model in a custom dataset?

### ISSUE THREAD

**Issue by chunniunai220ml** [qijiezhao/M2Det#30]
Thank for your sharing!
I want to konw how to prepare the custom train_val data, the data format is 
img_path x_min,y_min,x_max,y_max,cls ?

besides,how can i  finetune on your published model in a custom dataset?   or I'd better train from scratch

**Comment by qijiezhao** [comment]
Good question!
1) You can follow the data/voc.py to customize your dataloader.
2) It's OK to finetune from the published model, but you have to completely align the model size before loading the weight, you have to modify some code in utils/core.py.

**Comment by Roujack** [comment]
Hi, Thx for sharing your code. I've finetuned in your published model in my own dataset.I set the dataset in a VOC2007 format and I have set the num_class = 7 instead of 81. And I think relevant parameter are set properly
But the result is so bad! I test in your image and my own image, there are many bounding box(more than 1K).I am so confused. Could you tell me what's the problem?
![_COCO_train2014_000000000659](https://user-images.githubusercontent.com/20369575/54683339-607e5300-4b4c-11e9-905e-891ad7a98ef0.jpg)

**Comment by wanzn** [comment]
> Hi, Thx for sharing your code. I've finetuned in your published model in my own dataset.I set the dataset in a VOC2007 format and I have set the num_class = 7 instead of 81. And I think relevant parameter are set properly
> But the result is so bad! I test in your image and my own image, there are many bounding box(more than 1K).I am so confused. Could you tell me what's the problem?
> ![_COCO_train2014_000000000659](https://user-images.githubusercontent.com/20369575/54683339-607e5300-4b4c-11e9-905e-891ad7a98ef0.jpg)

i met the same problem, very confusing.

**Comment by dshahrokhian** [comment]
How do your losses look like during training? Did you check the results of `test.py` on some test data that you have?

**Comment by Roujack** [comment]
> Facing exact same issue.
> 
> How do your losses look like during training? Did you check the results of `test.py` on some test data that you have?

In my experiment，the console will print loss as default. The loss_L is about 3.237 after trainning and loss_C drop from 11 to about 3. Result(runing on test.py) on my test data is so bad too(so many bounding box)!

**Comment by MenGuangwen-CN-0411** [comment]
How about the test.py results?

**Comment by dshahrokhian** [comment]
Well, it seems that your training is still not converging, as those values are pretty high. In my case, I have `Loss_L = 1.28` and `Loss_C=0.4` but still lots of FP bounding boxes.

**Comment by qijiezhao** [comment]
Hi, sorry for late reply.
@Roujack, your loss values with only 7 categories are not so correct. For example, the stable VOC(with 20 categories) losses are about loss_l: 1.2~1.4, while loss_c: 0.3~0.5. @dshahrokhian may have a more stable training process.

The reason that have so many FP bboxes, I guess you can:
1) check the GT label: 1~k, not 0~k-1. There is a BG class.
2) Visualize the training images, it same as the val image, you should check the pre-process or post-process. If much better than the val images, you should check whether the training process is overfit. Because I don't how scale is your dataset, the training process needs to be tuned.

**Comment by MenGuangwen-CN-0411** [comment]
@qijiezhao ，Thanks for your sharing great object about objective detection. 
After 12 epochs on COCO, get the results as follow:
![_1](https://user-images.githubusercontent.com/48816239/54889525-04b32180-4ee0-11e9-9178-c21e3f3ab064.jpg)
![_5](https://user-images.githubusercontent.com/48816239/54889536-0da3f300-4ee0-11e9-8425-324479b3dd31.jpg)
About how many epochs you train your weights?

**Comment by qijiezhao** [comment]
> @qijiezhao ，Thanks for your sharing great object about objective detection.
> After 12 epochs on COCO, get the results as follow:
> ![_1](https://user-images.githubusercontent.com/48816239/54889525-04b32180-4ee0-11e9-9178-c21e3f3ab064.jpg)
> ![_5](https://user-images.githubusercontent.com/48816239/54889536-0da3f300-4ee0-11e9-8425-324479b3dd31.jpg)
> About how many epochs you train your weights?

Refer this: https://github.com/qijiezhao/M2Det/issues/30#issuecomment-476025270

**Comment by Roujack** [comment]
@qijiezhao problem is still unresolved==
here is some screenshots.
number of class setting:
<img width="383" alt="WechatIMG673" src="https://user-images.githubusercontent.com/20369575/54966983-f8e36000-4fb0-11e9-9708-37c0854e75a6.png">
<img width="690" alt="WechatIMG675" src="https://user-images.githubusercontent.com/20369575/54967016-0dbff380-4fb1-11e9-8f1a-c43d0088a063.png">
loss in 10 epoches(the dataset has 6464 images). It seems normal now:
![image](https://user-images.githubusercontent.com/20369575/54967071-46f86380-4fb1-11e9-8439-26af7a14d43c.png)
test on trainning image(so bad!):
![image](https://user-images.githubusercontent.com/20369575/54967138-8cb52c00-4fb1-11e9-93e4-5c98c0b24be3.png)

**Comment by MenGuangwen-CN-0411** [comment]
You shuold rewrite the dataloader part but not just modify as the above...

**Comment by dshahrokhian** [comment]
Are you using `pytorch==0.4.1` for both training and testing? That partially solved the problem for me, still trying to figure out how to improve results further.

**Comment by Roujack** [comment]
@dshahrokhian pytorch version is 0.4.1.

**Comment by dshahrokhian** [comment]
I see that you have a lot of bounding boxes with low confidences. Did you try increasing the threshold in `demo.py` function `draw_detection`? Something like 0.9.

**Comment by Roujack** [comment]
@dshahrokhian yeah. As I increase the score threshold with 0.9 in **demo.py** function draw_detection, and set the num_per_class = 5 in configs/m2det512_vgg.pth, the boundingbox number is decrease.
But the bounding box is not locate object well and the classification labels are wrong:
![6811553607441_ pic](https://user-images.githubusercontent.com/20369575/55001607-efd5ab80-500f-11e9-88ec-e80607787b36.jpg)
I train on my customized dataset in 10 epoch and the loss_l = 1.2 and loss_c = 0.3.
what do you think I can improve the detector behaviour?

**Comment by rasse-cep** [comment]
I met same issue.

**Comment by zhaobaozi** [comment]
hi,i am new in pytorch..how to finetuning base your model?Thanks

**Comment by zhaobaozi** [comment]
hi,when i finetuning,i got loss_l=0.6,and loss_c=1.2...the loss_c is big..
i have change the code about calss as your screenshots,have any where should be change?
Thanks @Roujack @qijiezhao

**Comment by DeqiangWang** [comment]
hi @dshahrokhian , i got loss_l=0.63,and loss_c=0.8, VOC0712 I can get a good result,mAP reach 82%, but when train my dateset(14000samples), the test result is very bad....do you think overfitting? or any suggset?

**Comment by chiputao0823** [comment]
> hi @dshahrokhian , i got loss_l=0.63,and loss_c=0.8, VOC0712 I can get a good result,mAP reach 82%, but when train my dateset(14000samples), the test result is very bad....do you think overfitting? or any suggse


> hi @dshahrokhian , i got loss_l=0.63,and loss_c=0.8, VOC0712 I can get a good result,mAP reach 82%, but when train my dateset(14000samples), the test result is very bad....do you think overfitting? or any suggset?

请问我在voc0712上跑了160个epoch,map只有65%，程序只改了batchsize,num_class还有VOC_CLASSES，想知道你具体怎么训练到82%的

**Comment by chiputao0823** [comment]
> > hi @dshahrokhian , i got loss_l=0.63,and loss_c=0.8, VOC0712 I can get a good result,mAP reach 82%, but when train my dateset(14000samples), the test result is very bad....do you think overfitting? or any suggse
> 
> > hi @dshahrokhian , i got loss_l=0.63,and loss_c=0.8, VOC0712 I can get a good result,mAP reach 82%, but when train my dateset(14000samples), the test result is very bad....do you think overfitting? or any suggset?
> 
> 请问我在voc0712上跑了160个epoch,map只有65%，程序只改了batchsize,num_class还有VOC_CLASSES，想知道你具体怎么训练到82%的



> hi @dshahrokhian , i got loss_l=0.63,and loss_c=0.8, VOC0712 I can get a good result,mAP reach 82%, but when train my dateset(14000samples), the test result is very bad....do you think overfitting? or any suggset?

用了coco的预训练吗

**Comment by qilicun** [comment]
Any progress here？

**Comment by xlm998** [comment]
@qijiezhao  
    I see this:  
   It's OK to finetune from the published model, but you have to completely align the model size before loading the weight, you have to modify some code in utils/core.py.
   How to modify the size ,and which codes have to  modify in utils/core.py,  would you please to help me? Thanks very much

**Comment by cherishfive** [comment]
> 嗨@dshahrokhian，我得了loss_l = 0.63，而loss_c = 0.8，VOC0712我可以得到一个好结果，mAP达到82％，但是当训练我的日期组（14000samples）时，测试结果非常糟糕....你觉得吗？过度拟合？或任何suggset？



Hello, I can only run 72.5% of mAP on the pascal voc2007 dataset. How do you adjust it?

**Comment by psyanglll** [comment]
I have the same issue of  so many  bounding box :(

**Comment by zhulei1228** [comment]
Has everybody solved the problem?My loss_l can drop but loss_c can't

**Comment by taehyunzzz** [comment]
@zhulei1228 I remember somebody mentioning weighting loss_l and loss_c. Maybe you should give it a try?

**Comment by devendraswamy** [comment]
> Thank for your sharing!
> I want to konw how to prepare the custom train_val data, the data format is
> img_path x_min,y_min,x_max,y_max,cls ?
> 
> besides,how can i finetune on your published model in a custom dataset? or I'd better train from scratch

did you modify the model for custom data ??

is it work fine ??
 
Thanking You !!
